src/test/scala/chiselTests/DataPrint.scala

```diff
@@ -0,0 +1,81 @@
+// See LICENSE for license details.
+
+package chiselTests
+
+import org.scalatest._
+
+import chisel3._
+import chisel3.experimental.{ChiselEnum, FixedPoint, RawModule, MultiIOModule}
+
+class DataPrintSpec extends ChiselFlatSpec with Matchers {
+  object EnumTest extends ChiselEnum {
+    val sNone, sOne, sTwo = Value
+  }
+
+  // TODO: dedup w/ BundleLiteralSpec
+  class BundleTest extends Bundle {
+    val a = UInt(8.W)
+    val b = Bool()
+
+    // Bundle literal constructor code, which will be auto-generated using macro annotations in
+    // the future.
+    import chisel3.core.BundleLitBinding
+    import chisel3.internal.firrtl.{ULit, Width}
+    // Full bundle literal constructor
+    def Lit(aVal: UInt, bVal: Bool): BundleTest = {
+      val clone = cloneType
+      clone.selfBind(BundleLitBinding(Map(
+        clone.a -> litArgOfBits(aVal),
+        clone.b -> litArgOfBits(bVal)
+      )))
+      clone
+    }
+  }
+
+  "Data types" should "have a meaningful string representation" in {
+    elaborate { new RawModule {
+      UInt().toString should be ("UInt")
+      UInt(8.W).toString should be ("UInt<8>")
+      SInt(15.W).toString should be ("SInt<15>")
+      Bool().toString should be ("Bool")
+      Clock().toString should be ("Clock")
+      FixedPoint(5.W, 3.BP).toString should be ("FixedPoint<5><<3>>")
+      Vec(3, UInt(2.W)).toString should be ("Vec<3, UInt<2>>")
+      EnumTest.Type().toString should be ("chiselTests.DataPrintSpec$EnumTest")
+      (new BundleTest).toString should be ("BundleTest")
+    } }
+  }
+
+  class BoundDataModule extends MultiIOModule {  // not in the test to avoid anon naming suffixes
+    Wire(UInt()).toString should be("UInt(Wire in DataPrintSpec$BoundDataModule)")
```

Comment:
Do you think it might be worth having a source locator for Data? Let people know where the Wire was constructed?

Reply:
Possibly. This is new enough that I don't know what the use case for this will be (other than printing literals), so I'd like to wait and see first.

There would also need to be infrastructure built for it, mainly IO needs to have a source locator attached to it, and all Bindings needs a copy of the source locator (currently is only part of the FIRRTL command). Adding source locators for literals would be more difficult because of the possibility for a chained apply, so we might have to build more source info transform macros.